bam_fastq.c

```diff
@@ -659,33 +333,58 @@ static bool parse_opts(int argc, char *argv[], bam2fq_opts_t** opts_out)
     return true;
 }
 
-static BGZF *open_fqfile(char *filename, int c, htsThreadPool *tp)
-{
-    char mode[4] = "w";
-    size_t len = strlen(filename);
-
-    mode[2] = 0; mode[3] = 0;
-    if (len > 3 && strstr(filename + (len - 3),".gz")) {
-        mode[1] = 'g'; mode[2] = c+'0';
-    } else if ((len > 4 && strstr(filename + (len - 4),".bgz"))
-               || (len > 5 && strstr(filename + (len - 5),".bgzf"))) {
-        mode[1] = c+'0';
-    } else {
-        mode[1] = 'u';
+void set_sam_opts(samFile *fp, bam2fq_state_t *state,
+                  const bam2fq_opts_t *opts) {
+    if (state->has12)
+        hts_set_opt(fp, FASTQ_OPT_RNUM, 1);
+
+    if (state->illumina_tag)
+        hts_set_opt(fp, FASTQ_OPT_CASAVA, 1);
+
+    hts_set_opt(fp, FASTQ_OPT_BARCODE, opts->barcode_tag);
+
+    kstring_t tag_list = {0,0};
+    if (state->copy_tags)
+        kputs("RG,BC,QT", &tag_list);
+    if (opts->extra_tags) {
+        if (tag_list.l)
+            kputc(',', &tag_list);
+        kputs(opts->extra_tags, &tag_list);
+    }
+    if (tag_list.l)
+        hts_set_opt(fp, FASTQ_OPT_AUX, tag_list.s);
+    ks_free(&tag_list);
+}
+
+// Open a file as normal or gzipped based on filename.
+// Note we always use bgzf and don't bother to attempt non-blocked
+// gzip streams.  This is a departure from the old fastq code.
+static samFile *sam_open_z(char *fn, char *mode, bam2fq_state_t *state) {
+    static char modez[6];
```

Comment:
Because you decreed at some polint that all functions in Samtools must be static.   https://github.com/samtools/samtools/pull/1093#discussion_r378271262. I remember discussing it in one of our meetings.

 I disagreed then, and still do, but this is why at the time I wanted such rules written down as I feared they'd be applied haphazzardly or flip-flop back and forth.

Reply:
I'm not referring to the function here.  `modez` is also static.

Reply:
Ugh my apologies.  I totally failed to notice that bit in the quoted source.

I've no idea why.  I'll investigate, but likely a mistake.